standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/dbinstall/AbstractDatabase.java

```diff
@@ -0,0 +1,302 @@
+  }
+
+  private int runCmdAndPrintStreams(String[] cmd, long secondsToWait)
+      throws InterruptedException, IOException {
+    ProcessResults results = runCmd(cmd, secondsToWait);
+    LOG.info("Stdout from proc: " + results.stdout);
+    LOG.info("Stderr from proc: " + results.stderr);
+    return results.rc;
+  }
+
+  private ProcessResults runCmd(String[] cmd, long secondsToWait)
+      throws IOException, InterruptedException {
+    LOG.info("Going to run: " + StringUtils.join(cmd, " "));
+    Process proc = Runtime.getRuntime().exec(cmd);
+    if (!proc.waitFor(secondsToWait, TimeUnit.SECONDS)) {
+      throw new RuntimeException(
+          "Process " + cmd[0] + " failed to run in " + secondsToWait + " seconds");
+    }
+    BufferedReader reader = new BufferedReader(new InputStreamReader(proc.getInputStream()));
+    final StringBuilder lines = new StringBuilder();
+    reader.lines().forEach(s -> lines.append(s).append('\n'));
+
+    reader = new BufferedReader(new InputStreamReader(proc.getErrorStream()));
+    final StringBuilder errLines = new StringBuilder();
+    reader.lines().forEach(s -> errLines.append(s).append('\n'));
+    LOG.info("Result size: " + lines.length() + ";" + errLines.length());
+    return new ProcessResults(lines.toString(), errLines.toString(), proc.exitValue());
+  }
+
+  public void launchDockerContainer() throws Exception {
+    runCmdAndPrintStreams(buildRmCmd(), 600);
+    if (runCmdAndPrintStreams(buildRunCmd(), 600) != 0) {
+      throw new RuntimeException("Unable to start docker container");
+    }
+    long startTime = System.currentTimeMillis();
+    ProcessResults pr;
+    do {
+      Thread.sleep(1000);
+      pr = runCmd(buildLogCmd(), 30);
+      if (pr.rc != 0) {
+        throw new RuntimeException("Failed to get docker logs");
+      }
+    } while (startTime + MAX_STARTUP_WAIT >= System.currentTimeMillis() && !isContainerReady(pr));
+    if (startTime + MAX_STARTUP_WAIT < System.currentTimeMillis()) {
+      throw new RuntimeException("Container failed to be ready in " + MAX_STARTUP_WAIT/1000 +
+          " seconds");
+    }
+  }
+
+  public void cleanupDockerContainer() throws IOException, InterruptedException {
+    if (runCmdAndPrintStreams(buildRmCmd(), 600) != 0) {
+      throw new RuntimeException("Unable to remove docker container");
+    }
+  }
+
+  @Override
+  public void before() throws Exception {
+    launchDockerContainer();
+    MetastoreSchemaTool.setHomeDirForTesting();
+  }
+
+  @Override
+  public void after() {
+    if ("true".equalsIgnoreCase(System.getProperty("metastore.itest.no.stop.container"))) {
+      LOG.warn("Not stopping container " + getDockerContainerName() + " at user request, please "
+          + "be sure to shut it down before rerunning the test.");
+      return;
+    }
+    try {
+      cleanupDockerContainer();
+    } catch (InterruptedException | IOException e) {
+      e.printStackTrace();
+    }
+  }
```

Comment:
Possibly these methods can go to those classes using this class as a rule. Another alternative would be to create a rule class which wraps an instance of `AbstractDatabase` and provides these methods.